### PR TITLE
Make extension components ready for cached client

### PR DESCRIPTION
**Pull request by timebertt**
**How to categorize this PR?**

/area dev-productivity scalability
/kind cleanup

**What this PR does / why we need it**:

Similar to previous PRs, this PR works towards getting rid of the `DirectClient`.
This one tackles usages in the extension components and makes them ready for cached clients.

**Which issue(s) this PR fixes**:
Part of #2822

**Special notes for your reviewer**:

c7bd0175eadbf34e8c4ac44eb4640c183d39fda3 is a prefactoring of the extension helper funcs, that is necessary for making the components ready for a cached client (see commit message for more details/reasoning).
The other commits perform the actual refactoring/adaption per component.

/squash

**Release note**:

```other operator

```

### CODE REVIEW

**Comment by timebertt** [comment]
@ialidzhikov, this is the bigger refactoring I was talking about earlier. 
Let's see how the review goes and if we have enough time left to validate it before the release is cut, we can include it, otherwise it doesn't hurt to wait a bit :)

**Comment by timebertt** [comment]
Strange...
The `verify` step failed twice with the following error: 
```
Summarizing 1 Failure:

[Fail] extensions #WaitUntilExtensionObjectReady [It] should set passed object to latest state once ready 
/go/src/github.com/gardener/gardener/pkg/extensions/customresources_test.go:214

Ran 40 of 40 Specs in 0.165 seconds
FAIL! -- 39 Passed | 1 Failed | 0 Pending | 0 Skipped
```

Though locally, in can't get the test to fail:
```
$ stress ./pkg/extensions/extensions.test
343 runs so far, 0 failures
696 runs so far, 0 failures
1016 runs so far, 0 failures
1320 runs so far, 0 failures
1603 runs so far, 0 failures
1891 runs so far, 0 failures
2166 runs so far, 0 failures
2441 runs so far, 0 failures
2702 runs so far, 0 failures
2946 runs so far, 0 failures
```

File: pkg/operation/botanist/botanist.go
Comment (rfranzke) [inline pkg/operation/botanist/botanist.go:71]: Unrelated to the PR, but as you are on it: Should we change the interface of all these `botanist.Default*` functions to not take a client? The `Default*` functions can access `b.K8sSeedClient` similarly, and already do for some other fields, so it's quite inconsistent atm.

File: pkg/operation/botanist/component/extensions/controlplane/controlplane.go
Comment (rfranzke) [inline pkg/operation/botanist/component/extensions/controlplane/controlplane.go]: Nit: You could still return `&controlPlane` right away, no need for the intermediate `c` variable

**Review by rfranzke — DISMISSED** [review]
/lgtm

Well done!

File: pkg/operation/botanist/botanist.go
Comment (timebertt) [reply pkg/operation/botanist/botanist.go:71]: Yep, sure. I'm already on the next PR and will clean this up there once I've removed some more of the `DirectClient` usages there.

**Comment by timebertt** [comment]
Thanks @rfranzke for your review, PTAL :)

**Comment by timebertt** [comment]
/hold
I think I found a bug.

**Comment by timebertt** [comment]
Indeed, I found a situation in which `MergePatchOrCreate` can be problematic, but I didn't notice until now, when switching from `CreateOrUpdate` to `MergePatchOrCreate` in this PR.

There's also one usage on `master` already, which needs to be fixed (introduced in https://github.com/gardener/gardener/pull/3991): https://github.com/gardener/gardener/blob/5589fcc84dfc76e8763cfa3ee44be5db57950e74/pkg/extensions/cluster.go#L93-L104

I'll fix the current usage on Friday and follow up with more details on the usages in this PR.

/needs changes
/author-action

**Comment by timebertt** [comment]
Some `MergePatchOrCreate` usages in this PR need to be changed, because of the reasons explained in https://github.com/gardener/gardener/pull/4057.
I will rebase the PR onto https://github.com/gardener/gardener/pull/4057 and adapt accordingly, once it's merged.

@ialidzhikov: as this PR is non-critical, I would propose to move it to the next release and not include it in `v1.23`, to not block the release any further.

**Comment by timebertt** [comment]
> Some `MergePatchOrCreate` usages in this PR need to be changed, because of the reasons explained in #4057.
> I will rebase the PR onto #4057 and adapt accordingly, once it's merged.

Rebased the PR and added two more commits to make it ready again:
- one to switch from `MergePatchOrCreate` to `GetAndCreateOrMergePatch`: 3576fd65abff9b73ec20dfbe98e82f4a7118165a
- and one to Remove the `*PatchOrCreate` funcs: 9cabff70f516d830bdca6fecfc5162bec18d57cc

/unassign
/unhold
/remove status/author-action

@rfranzke PTAL :)

**Review by rfranzke — APPROVED** [review]
/lgtm

**Comment by rfranzke** [comment]
@timebertt What about the `CreateOr{Strategic}MergePatch` function ([ref](https://github.com/gardener/gardener/search?q=CreateOrStrategicMergePatch))? Do you think their current usage is safe or should it be adapted as well (i.e., first try `Create`, but if the resource already exists, then we `Get` it and `Patch` it afterwards)?

**Comment by timebertt** [comment]
> @timebertt What about the `CreateOr{Strategic}MergePatch` function ([ref](https://github.com/gardener/gardener/search?q=CreateOrStrategicMergePatch))? Do you think their current usage is safe or should it be adapted as well (i.e., first try `Create`, but if the resource already exists, then we `Get` it and `Patch` it afterwards)?

Generally, the same applies to `CreateOr{Strategic}MergePatch` as to the `*PatchOrCreate` logic: if you're not trying to remove any optional fields in the patch, it's safe to use without doing a `Get` before. But if you do want to remove optional fields you have to fetch the object first, so that the diff contains an explicit `null` value.

So for example, this usage needs to be adapted, as labels that were removed from the seed template won't be removed from the seed itself: https://github.com/gardener/gardener/blob/93e4efbe30b45f8d9089b0c6e8131b1178b8e9ac/pkg/gardenlet/controller/managedseed/managedseed_actuator.go#L309

Probably, adapting the helper func to do a `Get` before `Patch` wouldn't hurt. But I haven't checked all the usages in detail. Though, such a helper func brings a good chance of misuse (that's why I removed the `*PatchOrCreate` funcs).

**Comment by rfranzke** [comment]
OK, thanks. For the sake of simplicity and prevention of coding bugs, I'd vote for changing the function in all cases as described above (create->get->patch) instead of (create->patch). Do you agree?

**Comment by timebertt** [comment]
Yes, I agree. That seems to be the more sustainable approach.

**Pull request merged**
